Add support for type checking Unstructured event objects via GVK to client-go's reflector

marun opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

It was previously possible to instantiate `Reflector` with `*unstructured.Unstructured` as the expected type but this did not
support checking that event objects were of the correct API type (e.g. if event object was `v1.Pod` in `Unstructured` form but
`v1.Service` was expected). This commit adds support for providing a GVK via an `Unstructured` expected type to compare with the GVK of event objects. The GVK will also be used in reflector log output.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/assign @sttts

sttts commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -49,7 +51,13 @@ type Reflector struct {
+	expectedTypeName string
+	// The GVK of the object we expect to place in the store.
+	// Set if the type is Unstructured.
+	expectedGVK *schema.GroupVersionKind
```
expectedUnstructuredGVK

marun replied:
Done

sttts commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -340,6 +365,12 @@ loop:
 				utilruntime.HandleError(fmt.Errorf("%s: expected type %v, but watch event object had type %v", r.name, e, a))
 				continue
 			}
+			if r.expectedGVK != nil {
```
would format the block above the same way:
```
if r.expectedType != nil {
   ...
}

marun replied:
Done

sttts commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -49,7 +51,13 @@ type Reflector struct {
 
 	// The type of object we expect to place in the store.
+	// Set if the type is not Unstructured.
 	expectedType reflect.Type
```
expectedNativeType

sttts replied:
And move the name on-top of this, so that the native and the unstructured fields are next to each other.

marun replied:
Done

sttts commented:
Left some nits. Looks good otherwise :+1:

sttts commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -102,22 +108,39 @@ func NewNamedReflector(name string, lw ListerWatcher, expectedType interface{},
 }
 
+func (r *Reflector) setExpectedType(expectedType interface{}) {
+	r.expectedTypeName = "<unknown>"
```
it's more like `<unspecified>` than unknown.

marun replied:
Done

fedebongio commented:
/assign @roycaihw

fedebongio commented:
/assign @jpbetz

jpbetz commented:
nit: Can we change the title? It makes it look like unstructured is not supported, but the issue being fixed is much more focused.

marun commented:
> nit: Can we change the title? It makes it look like unstructured is not supported, but the issue being fixed is much more focused.

Done

marun commented:
The failing `pull-kubernetes-bazel-test` was due to the dynamic informer test providing `&unstructured.Unstructured{}` as the expected type without populating GVK:

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go#L137

Since the dynamic informer is using `GroupVersionResource` (gvr), and I didn't see an easy way of passing gvr via the `Unstructured` expected type or retrieving gvr from the event object for comparison, I've updated `Reflector.setExpectedType` to use the reflected type if `expectedType` is `Unstructured` but does not have a populated `GroupVersionKind`.

liggitt commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -41,15 +43,21 @@ import (
 	metrics *reflectorMetrics
 
-	// The type of object we expect to place in the store.
-	expectedType reflect.Type
```
would it be simpler to keep the expectedType aspect as-is, and add an optional `expectedGVK *schema.GroupVersionKind` that we populate only in unstructured cases?

marun replied:
That was effectively what I originally proposed (at least in terms of variables) - add `expectedGVK` and `expectedTypeName` (to simplify logging of either either `expectedType.String()` or `expectedGVK.String()`. The renames to `expectedNativeType` and `expectedUnstructuredGVK` were requested by @sttts:

https://github.com/kubernetes/kubernetes/pull/82388#discussion_r321451910
https://github.com/kubernetes/kubernetes/pull/82388#discussion_r321450959

marun replied:
Apologies, I think I misinterpreted your intent. Would you prefer to leave the check of the reflected type as-is (always check if non-nil, even if unstructured), and add the check for gvk if provided by unstructured?

Regarding the logging, is it acceptable to log gvk instead of reflected type name if gvk were provided? Without this change, the log output of multiple informers with an unstructured expected type is undifferentiated.

liggitt replied:
> Would you prefer to leave the check of the reflected type as-is (always check if non-nil, even if unstructured), and add the check for gvk if provided by unstructured?

Yes, exactly.

> Regarding the logging, is it acceptable to log gvk instead of reflected type name if gvk were provided? Without this change, the log output of multiple informers with an unstructured expected type is undifferentiated.

Yes, improving the logging is fine.

marun replied:
I've updated the check for expected type to be effectively the same as it was previously. Would you prefer that I go further and revert to using `expectedType` instead of `expectedNativeType`?

liggitt replied:
> Would you prefer that I go further and revert to using expectedType instead of expectedNativeType?

I think so, since we check it in all cases

marun replied:
Done.

dims commented:
/uncc

liggitt commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -41,15 +43,21 @@ import (
 
-	// The type of object we expect to place in the store.
-	expectedType reflect.Type
+	// The name of the type we expect to place in the store.
```
clarify this is a stringification of expectedUnstructuredGVK if that is provided, and a stringification of expectedType otherwise

marun replied:
Done

liggitt commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -102,22 +110,43 @@ func NewNamedReflector(name string, lw ListerWatcher, expectedType interface{},
+		// Use gvk to check that watch event objects are of the desired type.
+		gvk := obj.GroupVersionKind()
+		if gvk.Empty() {
+			klog.V(4).Infof("Reflector from %s configured with expectedType of *unstructured.Unstructured with empty GroupVerionKind.", r.name)
```
```suggestion
			klog.V(4).Infof("Reflector from %s configured with expectedType of *unstructured.Unstructured with empty GroupVersionKind.", r.name)
```

marun replied:
Done

marun committed:
Add support for type checking Unstructured via GVK in reflector

It was previously possible to instantiate `Reflector` with
`*unstructured.Unstructured` as the expected type but this did not
support checking that event objects were of the correct API
type (e.g. if event object was `v1.Pod` in `Unstructured` form but
`v1.Service` was expected). This commit adds support for providing a
GVK via an `Unstructured` expected type to compare with the GVK of
event objects. The GVK will also be used in reflector log output.

marun committed:
Run hack/update-bazel.sh

marun commented:
Updated in response to review comments.

liggitt commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -100,22 +111,43 @@ func NewNamedReflector(name string, lw ListerWatcher, expectedType interface{},
 
+func (r *Reflector) setExpectedType(expectedType interface{}) {
+	if expectedType == nil {
+		r.expectedTypeName = defaultExpectedTypeName
```
still assign r.expectedType here?

marun replied:
Given that `reflect.TypeOf(nil) == nil`,  what would be the value of doing so?

marun replied:
Don't get me wrong, I'm happy to make the change.  I'd just like to understand the rational.

marun replied:
Done

liggitt replied:
> Don't get me wrong, I'm happy to make the change. I'd just like to understand the rational.

If setExpectedType was previously called, returning early here could leave r.expectedType set to the previous non-nil value

marun replied:
Ok, thank you.

marun commented:
/test pull-kubernetes-bazel-build

liggitt commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -331,9 +363,17 @@ loop:
+				}
+			}
+			if r.expectedGVK != nil {
+				if e, a := r.expectedGVK, event.Object.GetObjectKind().GroupVersionKind(); e.String() != a.String() {
```
sorry, just noticed this... is there a reason not to do this, rather than stringifying them?
```
if e, a := *r.expectedGVK, event.Object.GetObjectKind().GroupVersionKind(); e != a {
```

marun replied:
Thank you for catching my elementary mistake.  :)

liggitt commented:
verify failure:
```
reflector.go:56:48: "comparision" is a misspelling of "comparison"
```

marun commented:
A CI spell checker for comments, nice! I only realized afterwards that flyspell was telling me the same thing. :/

marun commented:
/test pull-kubernetes-typecheck

Job is marked as failing but the job details suggest it's still running.

liggitt commented:
/lgtm
/approve
/priority backlog

tedyu commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -331,9 +363,17 @@ loop:
+				}
+			}
+			if r.expectedGVK != nil {
+				if e, a := *r.expectedGVK, event.Object.GetObjectKind().GroupVersionKind(); e != a {
```
Nice
